Fix F58 encoding in non-multibyte locales

grondo opened the pull request:
As noted in #3142 and #3124, inclusion of the multibyte unicode ƒ in JOBIDs by default causes a serious usability problem in non-UTF-8 locales.

This PR attempts to adhere to the locale preference for `LC_CTYPE` by falling back to ascii "f" when [`MB_CUR_MAX(3)`](https://www.man7.org/linux/man-pages/man3/MB_CUR_MAX.3.html) returns 1.

The default locale for all programs at startup is `C`. In order to apply the current settings from the environment, a program must first call [`setlocale(3)`](https://www.man7.org/linux/man-pages/man3/setlocale.3.html) with an empty string (`setlocale (LC_CTYPE, "")`). Therefore, in order for the approach used in the PR to be effective, `MB_CUR_MAX` is checked, and if it is equal to `1`, `setlocale (LC_CTYPE, "")` is called and the result of `MB_CUR_MAX` is checked again.

Unfortunately, there does not seem to be way to determine if a program has already called `setlocale(3)` with a different locale, so there is a small chance that this approach will result in clobbering a locale that has been explicitly set by a program. I don't think that is very common, but it could result in confusing behavior.

Since the fluid functions could be called from a multi-threaded process (e.g. flux-broker), and `setlocale(3)` is explicitly MT-unsafe, we have to do the locale checking under a mutex. Fortunately, this only has to be done once per process (at least for now, we don't support changing the locale from the FLUID F58 encoding perspective, as that is outside the scope of this fix)

Hopefully all of that makes sense. The `locale_is_utf8()` function should get some extra scrutiny by @garlick. (Never sure I get the pthreads stuff right anymore)

```
ƒ(s=1,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux jobs -a -c 10
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME RANKS
    ƒGCjPrwu grondo   sleep      CD      1      1   0.091s 0
    ƒGCjPrwt grondo   sleep      CD      1      1   0.125s 0
    ƒGCjPrws grondo   sleep      CD      1      1   0.137s 0
    ƒGCjPrwr grondo   sleep      CD      1      1   0.144s 0
    ƒGCjPrwq grondo   sleep      CD      1      1   0.146s 0
    ƒGChusfZ grondo   sleep      CD      1      1   0.146s 0
    ƒGChusfY grondo   sleep      CD      1      1   0.151s 0
    ƒGChusfX grondo   sleep      CD      1      1   0.154s 0
    ƒGChusfW grondo   sleep      CD      1      1   0.142s 0
    ƒGChusfV grondo   sleep      CD      1      1   0.147s 0
ƒ(s=1,d=0,builddir) grondo@asp:~/git/flux-core.git$ LC_CTYPE=C flux jobs -a -c 10
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME RANKS
    fGCjPrwu grondo   sleep      CD      1      1   0.091s 0
    fGCjPrwt grondo   sleep      CD      1      1   0.125s 0
    fGCjPrws grondo   sleep      CD      1      1   0.137s 0
    fGCjPrwr grondo   sleep      CD      1      1   0.144s 0
    fGCjPrwq grondo   sleep      CD      1      1   0.146s 0
    fGChusfZ grondo   sleep      CD      1      1   0.146s 0
    fGChusfY grondo   sleep      CD      1      1   0.151s 0
    fGChusfX grondo   sleep      CD      1      1   0.154s 0
    fGChusfW grondo   sleep      CD      1      1   0.142s 0
    fGChusfV grondo   sleep      CD      1      1   0.147s 0
ƒ(s=1,d=0,builddir) grondo@asp:~/git/flux-core.git$ LC_ALL=C flux jobs -a -c 10
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME RANKS
    fGCjPrwu grondo   sleep      CD      1      1   0.091s 0
    fGCjPrwt grondo   sleep      CD      1      1   0.125s 0
    fGCjPrws grondo   sleep      CD      1      1   0.137s 0
    fGCjPrwr grondo   sleep      CD      1      1   0.144s 0
    fGCjPrwq grondo   sleep      CD      1      1   0.146s 0
    fGChusfZ grondo   sleep      CD      1      1   0.146s 0
    fGChusfY grondo   sleep      CD      1      1   0.151s 0
    fGChusfX grondo   sleep      CD      1      1   0.154s 0
    fGChusfW grondo   sleep      CD      1      1   0.142s 0
    fGChusfV grondo   sleep      CD      1      1   0.147s 0
```

grondo commented on src/common/libutil/fluid.c:
```diff
@@ -169,27 +170,68 @@ static int b58revenc (char *buf, int bufsz, fluid_t id)
+    return (MB_CUR_MAX > 1);
+}
+
+static int locale_is_utf8 (void)
```
Good thought! Honestly I had attempted this fix so many different ways that I think I lost sight of the forest for the trees.

> if we end up using wide characters in other places besides fluid_f58_encode() are we going to regret burying the locale initialization in the fluid class? 

Yeah, from what I know, it does seem a bit wrong to call `setlocale(3)` down in the depths of a library call. However, `setlocale(LC_CTYPE, "")` should be idempotent afaik. This would, however, step on any program that explicitly chooses a different locale than the environment or system defaults dictate.

> Would it be slightly more straightforward to just document that users of fluid_f58_encode() should be sure to call setlocale(LC_TYPE, "") in their main program initialization?

I wouldn't call it more straightforward, since we'd have to figure out where to document the instructions (`fluid_f58_encode()` is not a public function), and the documentation has to be added to both the C and Python bindings. That being said, it seems like the right thing to do.

Maybe Python already initializes the locale for all I know.

I'll work on a fix for this. It will make the changes in `libutil/fluid.c` trivial at least.

Thanks!

grondo replied:
[This](https://docs.python.org/3/library/locale.html#background-details-hints-tips-and-caveats) section of the Python `locale` documentation backs up what you are saying @garlick:

> It is generally a bad idea to call setlocale() in some library routine, since as a side effect it affects the entire program. Saving and restoring it is almost as bad: it is expensive and affects other threads that happen to run before the settings have been restored.

I am very glad you spoke up in this PR.

It also sounds like with Python at least, `LC_CTYPE` is automatically set from the environment by default, so no need to initialize locale when using Python bindings (which is the majority of use cases currently)

```
$ python3 -c 'import locale; print (locale.getlocale())'
('en_US', 'UTF-8')
```

garlick replied:
Hey great, so `flux jobs` should just work once `fluid_f58_encode()` respects `MB_CUR_MAX`?

grondo replied:
Yes. That looks to be the case.

grondo commented:
Ok, I've reworked this PR to use the approach suggested by @garlick above.

The changes to `fluid_f58_encode(3)` are now minimal, the function simply checks `MB_CUR_MAX` to respect the current LC_CTYPE encoding.

Additionally, calls to `setlocale (LC_ALL, "")` are now added to `flux-job`, `flux-shell` as well as the necessary unit tests.

Finally, since the output of `flux_job_id_encode(3)` is now dependent on the current locale, the libjob unit test is made more resilient to locale changes.

I'll want to test this on TOSS 3 clusters before a merge.

I had to rebase and squash along the way, so this PR needs a fresh look (though it should be simpler to review).

Looks like Python works as we'd hoped:
```
bash-4.4$ LC_ALL=C flux job id --to=f58 123
f38
bash-4.4$ LC_ALL=C.UTF-8 flux job id --to=f58 123
ƒ38
bash-4.4$ flux jobs -a 
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME RANKS
    ƒCpKX1YB grondo   hostname   CD      1      1   0.054s 0
    ƒA1a9rGf grondo   hostname   CD      1      1   0.059s 0
    ƒ6BEyZ7M grondo   hostname   CD      1      1   0.057s 0
    ƒ3sjeEEK grondo   hostname   CD      1      1   0.059s 0
    ƒ3Q43e2s grondo   hostname   CD      1      1   0.209s 0
bash-4.4$ LC_ALL=C flux jobs -a
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME RANKS
    fCpKX1YB grondo   hostname   CD      1      1   0.054s 0
    fA1a9rGf grondo   hostname   CD      1      1   0.059s 0
    f6BEyZ7M grondo   hostname   CD      1      1   0.057s 0
    f3sjeEEK grondo   hostname   CD      1      1   0.059s 0
    f3Q43e2s grondo   hostname   CD      1      1   0.209s 0
```

grondo commented:
Still having problems with the `python/t0010-job.py` test. It currently hardcodes the unicode prefix for FLUID in the F58 test and therefore fails if the current locale is C/POSIX.

grondo commented:
Tried various approaches to fix `python/t0010-job.py` but in the end I've just skipped the f58 jobid tests when the current encoding is "ascii".

garlick commented:
Looks good - I already gave my approval, so feel free to set MWP once you're ready.

grondo commented:
Thanks! BTW, I did try this out on TOSS 3 with `LC_ALL=C` and it seemed to work as expected (normal ascii `f` in output)
